tracing-attributes/src/lib.rs

```diff
@@ -835,125 +893,155 @@ mod kw {
+    // of the block, so that if we are working on a function whose
+    // `trait` or `impl` declaration is annotated by async_trait,
+    // this is quite likely the point where the future is pinned)
+    let (last_expr_stmt, last_expr) = block.stmts.iter().rev().find_map(|stmt| {
+        if let Stmt::Expr(expr) = stmt {
+            Some((stmt, expr))
+        } else {
+            None
         }
+    })?;
+
+    // is the last expression a function call?
+    let (outside_func, outside_args) = match last_expr {
+        Expr::Call(ExprCall { func, args, .. }) => (func, args),
+        _ => return None,
+    };
+
+    // is it a call to `Box::pin()`?
+    const BOX_PIN_NAMES: [&str; 6] = [
+        "Box::pin",
+        "std::boxed::Box::pin",
+        "::std::boxed::Box::pin",
+        "alloc::boxed::Box::pin",
+        "::alloc::boxed::Box::pin",
+        "boxed::Box::pin",
+    ];
+    let path = match outside_func.as_ref() {
+        Expr::Path(path) => &path.path,
+        _ => return None,
+    };
+    if !BOX_PIN_NAMES.contains(&path_to_string(path).as_str()) {
+        return None;
     }
-    None
-}
 
-struct AsyncTraitInfo {
-    name: String,
-    self_type: Option<syn::TypePath>,
-}
+    // Does the call takes an argument? If it doesn't,
+    // it's not gonna compile anyway, but that's no reason
+    // to (try to) perform an out of bounds access
+    if outside_args.is_empty() {
+        return None;
+    }
+
+    // Is the argument to Box::pin an async block that
+    // captures its arguments?
+    if let Expr::Async(async_expr) = &outside_args[0] {
+        // check that the move 'keyword' is present
+        async_expr.capture?;
 
-// Return the informations necessary to process a function annotated with async-trait.
-fn get_async_trait_info(block: &Block, block_is_async: bool) -> Option<AsyncTraitInfo> {
-    let fun = get_async_trait_function(block, block_is_async)?;
+        return Some(AsyncTraitInfo {
+            source_stmt: last_expr_stmt,
+            kind: AsyncTraitKind::Async(async_expr),
+            self_type: None,
+        });
+    }
+
+    // Is the argument to Box::pin a function call itself?
+    let func = match &outside_args[0] {
+        Expr::Call(ExprCall { func, .. }) => func,
+        _ => return None,
+    };
 
-    // if "_self" is present as an argument, we store its type to be able to rewrite "Self" (the
+    // "stringify" the path of the function called
+    let func_name = match **func {
+        Expr::Path(ref func_path) => path_to_string(&func_path.path),
+        _ => return None,
+    };
+
+    // Was that function defined inside of the current block?
+    // If so, retrieve the statement where it was declared and the function itself
+    let (stmt_func_declaration, func) = match inside_funs
+        .into_iter()
+        .find(|(_, fun)| fun.sig.ident == func_name)
+    {
+        Some((stmt, fun)) => (stmt, fun),
+        None => return None,
+    };
```

Comment:
nit, take it or leave it: i think this can be simplified to just
```suggestion
    let (stmt_func_declaration, func) =  inside_funs
        .into_iter()
        .find(|(_, fun)| fun.sig.ident == func_name)?;
```